import os
import glob
import pandas as pd
from snakemake.utils import min_version

min_version('5.26.1')

#container: 'continuumio/miniconda3:4.9.1'

configfile: '../config/hpcnode.yaml'

SAMPLES = pd.read_table(config['samples'])['Sample'].tolist()
CHROMOSOMES = pd.read_table(config['chromosomes'], header=None).iloc[:,0].tolist()

TMPDIR = config['paths']['temp_dir']
RAW_READ_DIR = '{0}/raw_data'.format(config['data_prefix'])
TRIMMED_READ_DIR = '{0}/trimmed_reads'.format(config['data_prefix'])
QC_DIR = '{0}/qc'.format(config['results_prefix'])
BAM_DIR = '{0}/bam'.format(config['results_prefix'])
VARIANT_DIR = '{0}/variants'.format(config['results_prefix'])
PROGRAM_RESOURCE_DIR = '{0}/program_resources'.format(config['results_prefix'])
ANGSD_DIR = '{0}/angsd'.format(config['results_prefix'])
NGSLD_DIR = '{0}/ngsld'.format(config['results_prefix'])
POP_STRUC_DIR = '{0}/population_structure'.format(config['results_prefix'])
FIGURES_DIR = '{0}/figures'.format(config['results_prefix'])

REFERENCE_GENOME = config['files']['reference_genome']
GFF_FILE = config['files']['gff']

REPRESENTATIVE_SAMPLE = config['variant_calling']['sample']
NODES_PER_CHROM = config['variant_calling']['nodes_per_chrom']
CORES_PER_NODE = config['variant_calling']['cores_per_node']
NODES = ['node{0:0=2d}'.format(x + 1) for x in range(NODES_PER_CHROM)]
if NODES_PER_CHROM == 1:
    NUM_REGIONS_PER_CHROM = CORES_PER_NODE
elif NODES_PER_CHROM > 1:
    NUM_REGIONS_PER_CHROM = NODES_PER_CHROM * CORES_PER_NODE

NGSADMIX_K = [x for x in range(1, 11)]
NGSADMIX_SEEDS = [x for x in range(1, 11)]

wildcard_constraints:
    chrom='|'.join([x for x in CHROMOSOMES])

localrules: create_tmp_dir, create_bam_list, region_files_forFreebayes, variant_calling_done, concat_angsd_stats, create_pos_file_for_ngsLD, concat_sfs, sum_sfs, convert_sites_for_angsd, split_angsd_sites_byChrom, angsd_index_sites, files_for_angsd_site_subset, subset_angsd_gl, subset_angsd_maf, extract_sample_angsd, angsd_done

include: 'rules/common.smk'

#raw_read_dict = create_raw_read_dict(RAW_READ_DIR, SAMPLES)

rule all:
   input:
        # Trimming, mapping, and QC
        expand('{0}/plastid_genes/{{gene}}/{{sample}}_{{gene}}.bam'.format(BAM_DIR), gene=['rbcl','matk'], sample=SAMPLES),
        '{0}/multiqc/multiqc_report.html'.format(QC_DIR), 
        # Variant calling
        '{0}/variant_calling.done'.format(VARIANT_DIR),
        # Angsd
        '{0}/angsd.done'.format(ANGSD_DIR),
        # Population structure
        expand('{0}/pcangsd/allSamples_{{site}}_maf{{maf}}_pcangsd.{{ext}}'.format(POP_STRUC_DIR), site=['4fold', '0fold'], ext=['cov', 'admix.Q.npy'], maf=['0.05']),
        expand('{0}/ngsadmix/K{{k}}/allSamples_ngsadmix_{{site}}_maf{{maf}}_K{{k}}_seed{{seed}}.fopt.gz'.format(POP_STRUC_DIR), k=NGSADMIX_K, seed=NGSADMIX_SEEDS, site=['4fold'], maf=['0.05']),
        # Param tests
        expand('{0}/test_params/GL{{GL}}_baq{{baq}}/CM019112.1_allSamples_allSites_GL{{GL}}_baq{{baq}}.saf.{{ext}}'.format(ANGSD_DIR), chrom=CHROMOSOMES, ext=['gz','idx','pos.gz'], GL=['1','2'], baq=['0','1','2'])
        # NGSLD
        #expand('{0}/pruned/{{chrom}}/{{chrom}}_withMaf{{maf}}_pruned.id'.format(NGSLD_DIR), chrom=CHROMOSOMES, maf=['0.05']),

rule create_tmp_dir:
    output: directory(TMPDIR)
    shell: 'mkdir {output}'

include: 'rules/ref.smk'
include: 'rules/trimming.smk'
include: 'rules/mapping.smk'
include: 'rules/qc.smk'
include: 'rules/variant_calling.smk'
include: 'rules/param_tests.smk'
include: 'rules/angsd.smk'
include: 'rules/population_structure.smk'

rule clean:
    params:
        'logs {0} {1}/* {2} fastp.json'.format(config['results_prefix'], TRIMMED_READ_DIR, TMPDIR)
    shell:
        'rm -rfv {params}'
